Highlighting error on raw string literal

### Description
Highlighting error on raw string literal [log-snippet.txt](https://github.com/SonarOpenCommunity/sonar-cxx/files/3677049/log-snippet.txt)

### Steps to reproduce the problem
Run sonar-scanner with the attached example code [main.cpp.txt](https://github.com/SonarOpenCommunity/sonar-cxx/files/3677062/main.cpp.txt)

### Related information
* cxx plugin version: 1.3.1
* SonarQube version: 7.9.1

SonarQube version: 7.9.1
cxx_plugin version: 1.3.1

I have a similar error, but only in Docker Container.

`
13:03:38.199 DEBUG: highlighting exception {}
java.lang.IllegalArgumentException: Unable to highlight file src/accept.cpp
	at org.sonar.api.batch.sensor.highlighting.internal.DefaultHighlighting.highlight(DefaultHighlighting.java:97)
	at org.sonar.api.batch.sensor.highlighting.internal.DefaultHighlighting.highlight(DefaultHighlighting.java:37)
	at org.sonar.cxx.sensors.visitors.CxxHighlighterVisitor.highlight(CxxHighlighterVisitor.java:113)
	at org.sonar.cxx.sensors.visitors.CxxHighlighterVisitor.visitToken(CxxHighlighterVisitor.java:96)
	at com.sonar.sslr.impl.ast.AstWalker.visitToken(AstWalker.java:107)
	at com.sonar.sslr.impl.ast.AstWalker.visit(AstWalker.java:86)
	at com.sonar.sslr.impl.ast.AstWalker.visitChildren(AstWalker.java:99)
	at com.sonar.sslr.impl.ast.AstWalker.visit(AstWalker.java:87)
	at com.sonar.sslr.impl.ast.AstWalker.visitChildren(AstWalker.java:99)
	at com.sonar.sslr.impl.ast.AstWalker.visit(AstWalker.java:87)
	at com.sonar.sslr.impl.ast.AstWalker.visitChildren(AstWalker.java:99)
	at com.sonar.sslr.impl.ast.AstWalker.visit(AstWalker.java:87)
	at com.sonar.sslr.impl.ast.AstWalker.visitChildren(AstWalker.java:99)
	at com.sonar.sslr.impl.ast.AstWalker.visit(AstWalker.java:87)
	at com.sonar.sslr.impl.ast.AstWalker.visitChildren(AstWalker.java:99)
	at com.sonar.sslr.impl.ast.AstWalker.visit(AstWalker.java:87)
	at com.sonar.sslr.impl.ast.AstWalker.walkAndVisit(AstWalker.java:69)
	at org.sonar.squidbridge.AstScanner.scanFiles(AstScanner.java:110)
	at org.sonar.cxx.sensors.squid.CxxSquidSensor.execute(CxxSquidSensor.java:162)
	at org.sonar.scanner.sensor.AbstractSensorWrapper.analyse(AbstractSensorWrapper.java:48)
	at org.sonar.scanner.sensor.ModuleSensorsExecutor.execute(ModuleSensorsExecutor.java:85)
	at org.sonar.scanner.sensor.ModuleSensorsExecutor.lambda$execute$1(ModuleSensorsExecutor.java:59)
	at org.sonar.scanner.sensor.ModuleSensorsExecutor.withModuleStrategy(ModuleSensorsExecutor.java:77)
	at org.sonar.scanner.sensor.ModuleSensorsExecutor.execute(ModuleSensorsExecutor.java:59)
	at org.sonar.scanner.scan.ModuleScanContainer.doAfterStart(ModuleScanContainer.java:82)
	at org.sonar.core.platform.ComponentContainer.startComponents(ComponentContainer.java:136)
	at org.sonar.core.platform.ComponentContainer.execute(ComponentContainer.java:122)
	at org.sonar.scanner.scan.ProjectScanContainer.scan(ProjectScanContainer.java:400)
	at org.sonar.scanner.scan.ProjectScanContainer.scanRecursively(ProjectScanContainer.java:395)
	at org.sonar.scanner.scan.ProjectScanContainer.doAfterStart(ProjectScanContainer.java:358)
	at org.sonar.core.platform.ComponentContainer.startComponents(ComponentContainer.java:136)
	at org.sonar.core.platform.ComponentContainer.execute(ComponentContainer.java:122)
	at org.sonar.scanner.bootstrap.GlobalContainer.doAfterStart(GlobalContainer.java:141)
	at org.sonar.core.platform.ComponentContainer.startComponents(ComponentContainer.java:136)
	at org.sonar.core.platform.ComponentContainer.execute(ComponentContainer.java:122)
	at org.sonar.batch.bootstrapper.Batch.doExecute(Batch.java:73)
	at org.sonar.batch.bootstrapper.Batch.execute(Batch.java:67)
	at org.sonarsource.scanner.api.internal.batch.BatchIsolatedLauncher.execute(BatchIsolatedLauncher.java:46)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.base/java.lang.reflect.Method.invoke(Unknown Source)
	at org.sonarsource.scanner.api.internal.IsolatedLauncherProxy.invoke(IsolatedLauncherProxy.java:60)
	at com.sun.proxy.$Proxy0.execute(Unknown Source)
	at org.sonarsource.scanner.api.EmbeddedScanner.doExecute(EmbeddedScanner.java:189)
	at org.sonarsource.scanner.api.EmbeddedScanner.execute(EmbeddedScanner.java:138)
	at org.sonarsource.scanner.cli.Main.execute(Main.java:112)
	at org.sonarsource.scanner.cli.Main.execute(Main.java:75)
	at org.sonarsource.scanner.cli.Main.main(Main.java:61)
Caused by: java.lang.IllegalArgumentException: 128 is not a valid line offset for pointer. File src/accept.cpp has 70 character(s) at line 34
	at org.sonar.api.internal.google.common.base.Preconditions.checkArgument(Preconditions.java:145)
	at org.sonar.api.batch.fs.internal.DefaultInputFile.checkValid(DefaultInputFile.java:339)
	at org.sonar.api.batch.fs.internal.DefaultInputFile.newPointer(DefaultInputFile.java:272)
	at org.sonar.api.batch.fs.internal.DefaultInputFile.newRange(DefaultInputFile.java:288)
	at org.sonar.api.batch.sensor.highlighting.internal.DefaultHighlighting.highlight(DefaultHighlighting.java:95)
	... 48 common frames omitted
`

On host error does not appear ¯\_(ツ)_/¯

--- a/cxx-checks/pom.xml
+++ b/cxx-checks/pom.xml
@@ -5,7 +5,7 @@
   <parent>
     <groupId>org.sonarsource.sonarqube-plugins.cxx</groupId>
     <artifactId>cxx</artifactId>
-    <version>1.3.0-SNAPSHOT</version>
+    <version>1.3.2-SNAPSHOT</version>
   </parent>
 
   <artifactId>cxx-checks</artifactId>

--- a/cxx-sensors/pom.xml
+++ b/cxx-sensors/pom.xml
@@ -5,7 +5,7 @@
   <parent>
     <groupId>org.sonarsource.sonarqube-plugins.cxx</groupId>
     <artifactId>cxx</artifactId>
-    <version>1.3.0-SNAPSHOT</version>
+    <version>1.3.2-SNAPSHOT</version>
   </parent>
 
   <artifactId>cxx-sensors</artifactId>

--- a/cxx-sensors/src/main/java/org/sonar/cxx/sensors/visitors/CxxHighlighterVisitor.java
+++ b/cxx-sensors/src/main/java/org/sonar/cxx/sensors/visitors/CxxHighlighterVisitor.java
@@ -54,7 +54,7 @@ public class CxxHighlighterVisitor extends SquidAstVisitor<Grammar> implements A
   public void visitFile(@Nullable AstNode astNode) {
     newHighlighting = context.newHighlighting();
     InputFile inputFile = context.fileSystem().inputFile(context.fileSystem().predicates()
-      .is(getContext().getFile().getAbsoluteFile()));
+            .is(getContext().getFile().getAbsoluteFile()));
     if (inputFile != null) {
       newHighlighting.onFile(inputFile);
     }
@@ -83,10 +83,10 @@ public class CxxHighlighterVisitor extends SquidAstVisitor<Grammar> implements A
       } else if (token.getType().equals(CxxTokenType.STRING)) {
         Optional<Trivia> triviaWithConcatenatedLiterals = getTriviaWithConcatenatedLiterals(token);
         if (!triviaWithConcatenatedLiterals.isPresent()) {
-          last = highlight(last, new TokenLocation(token), TypeOfText.STRING);
+          last = highlight(last, new StringLocation(token), TypeOfText.STRING);
         } else {
           for (Token concatenatedLiterals : triviaWithConcatenatedLiterals.get().getTokens()) {
-            last = highlight(last, new TokenLocation(concatenatedLiterals), TypeOfText.STRING);
+            last = highlight(last, new StringLocation(concatenatedLiterals), TypeOfText.STRING);
           }
         }
       }
@@ -95,7 +95,7 @@ public class CxxHighlighterVisitor extends SquidAstVisitor<Grammar> implements A
         if (trivia.isComment()) {
           highlight(last, new CommentLocation(trivia.getToken()), TypeOfText.COMMENT);
         } else if (trivia.isSkippedText()
-          && trivia.getToken().getType().equals(CxxTokenType.PREPROCESSOR)) {
+                && trivia.getToken().getType().equals(CxxTokenType.PREPROCESSOR)) {
           highlight(last, new PreprocessorDirectiveLocation(trivia.getToken()), TypeOfText.PREPROCESS_DIRECTIVE);
         }
       }
@@ -104,20 +104,24 @@ public class CxxHighlighterVisitor extends SquidAstVisitor<Grammar> implements A
 
   private Optional<Trivia> getTriviaWithConcatenatedLiterals(Token stringToken) {
     return stringToken.getTrivia().stream()
-      .filter(t -> t.isSkippedText() && CxxTokenType.STRING.equals(t.getToken().getType())).findFirst();
+            .filter(t -> t.isSkippedText() && CxxTokenType.STRING.equals(t.getToken().getType())).findFirst();
   }
 
   private TokenLocation highlight(TokenLocation last, TokenLocation current, TypeOfText typeOfText) {
     try {
       if (!current.overlaps(last)) {
         newHighlighting.highlight(current.startLine(), current.startLineOffset(),
-          current.endLine(), current.endLineOffset(), typeOfText);
+                current.endLine(), current.endLineOffset(), typeOfText);
       }
     } catch (IllegalArgumentException ex) {
       // ignore highlight errors: parsing errors could lead to wrong location data
-      LOG.warn("Highlighting error in file '{}' at line:{}, column:{}", getContext().getFile().getAbsoluteFile(),
-        current.startLine(), current.startLineOffset());
-      LOG.debug("highlighting exception {}", ex);
+      LOG.debug("Highlighting error in file '{}' at start:{}:{} end:{}:{}",
+              getContext().getFile().getAbsoluteFile(),
+              current.startLine(),
+              current.startLineOffset(),
+              current.endLine(),
+              current.endLineOffset()
+      );
     }
     return current;
   }
@@ -155,15 +159,31 @@ public class CxxHighlighterVisitor extends SquidAstVisitor<Grammar> implements A
     public boolean overlaps(@Nullable TokenLocation other) {
       if (other != null) {
         return !(startLineOffset() > other.endLineOffset()
-          || other.startLineOffset() > endLineOffset()
-          || startLine() > other.endLine()
-          || other.startLine() > endLine());
+                || other.startLineOffset() > endLineOffset()
+                || startLine() > other.endLine()
+                || other.startLine() > endLine());
       }
       return false;
     }
 
   }
 
+  private static class StringLocation extends TokenLocation {
+
+    public StringLocation(Token token) {
+      super(token);
+      String value = token.getValue();
+      if (value.startsWith("R")) { // Raw String?
+        String[] lines = CxxUtils.EOL_PATTERN.split(value, -1);
+
+        if (lines.length > 1) {
+          endLine = token.getLine() + lines.length - 1;
+          endLineOffset = lines[lines.length - 1].length();
+        }
+      }
+    }
+  }
+  
   private static class CommentLocation extends TokenLocation {
 
     public CommentLocation(Token token) {

--- a/cxx-sensors/src/test/java/org/sonar/cxx/sensors/visitors/CxxHighlighterTest.java
+++ b/cxx-sensors/src/test/java/org/sonar/cxx/sensors/visitors/CxxHighlighterTest.java
@@ -47,7 +47,7 @@ public class CxxHighlighterTest {
 
     String content = Files.contentOf(target, StandardCharsets.UTF_8);
     DefaultInputFile inputFile = TestInputFileBuilder.create("ProjectKey", baseDir, target)
-      .setContents(content).setCharset(StandardCharsets.UTF_8).build();
+            .setContents(content).setCharset(StandardCharsets.UTF_8).build();
 
     context = SensorContextTester.create(baseDir);
     context.fileSystem().add(inputFile);
@@ -113,6 +113,9 @@ public class CxxHighlighterTest {
     checkOnRange(96, 25, 7, TypeOfText.STRING);
     checkOnRange(97, 25, 6, TypeOfText.STRING);
     checkOnRange(98, 25, 5, TypeOfText.STRING);
+    
+    // R"(\r\n      Hello World!\r\n   )" issue #1768
+    check(103, 14, TypeOfText.STRING);
   }
 
   @Test
@@ -127,20 +130,17 @@ public class CxxHighlighterTest {
   @SuppressWarnings("squid:S2699") // ... checkOnRange contains the assertion
   public void comment() {
 
-    check(1, 0, TypeOfText.COMMENT);
-    /*\r\n comment\r\n*/
+    check(1, 0, TypeOfText.COMMENT); /*\r\n comment\r\n*/
     check(3, 1, TypeOfText.COMMENT);
 
-    checkOnRange(5, 0, 2, TypeOfText.COMMENT);   //
-    checkOnRange(6, 0, 10, TypeOfText.COMMENT);  // comment
-    checkOnRange(7, 0, 2, TypeOfText.COMMENT);   //
+    checkOnRange(5, 0, 2, TypeOfText.COMMENT);    //
+    checkOnRange(6, 0, 10, TypeOfText.COMMENT);   // comment
+    checkOnRange(7, 0, 2, TypeOfText.COMMENT);    //
 
     checkOnRange(57, 22, 10, TypeOfText.COMMENT); // comment
     checkOnRange(58, 3, 10, TypeOfText.COMMENT);  // comment
-    checkOnRange(61, 3, 13, TypeOfText.COMMENT);
-    /* comment */
-    checkOnRange(64, 20, 13, TypeOfText.COMMENT);
-    /* comment */
+    checkOnRange(61, 3, 13, TypeOfText.COMMENT);  /* comment */
+    checkOnRange(64, 20, 13, TypeOfText.COMMENT); /* comment */
   }
 
   @Test
@@ -180,8 +180,8 @@ public class CxxHighlighterTest {
   }
 
   /**
-   * Checks the highlighting of a range of columns. The first column of a line has index 0. The range is the columns of
-   * the token.
+   * Checks the highlighting of a range of columns. The first column of a line
+   * has index 0. The range is the columns of the token.
    */
   private void checkOnRange(int line, int firstColumn, int length, TypeOfText expectedTypeOfText) {
     // check that every column of the token is highlighted (and with the expected type)

--- a/cxx-sensors/src/test/resources/org/sonar/cxx/sensors/highlighter.cc
+++ b/cxx-sensors/src/test/resources/org/sonar/cxx/sensors/highlighter.cc
@@ -98,4 +98,11 @@ void test3()
                          "Lou";
 }
 
+void test4()
+{
+   auto txt = R"(
+      Hello World!
+   )"; 
+}
+
 /* EOF */

--- a/cxx-squid/pom.xml
+++ b/cxx-squid/pom.xml
@@ -5,7 +5,7 @@
   <parent>
     <groupId>org.sonarsource.sonarqube-plugins.cxx</groupId>
     <artifactId>cxx</artifactId>
-    <version>1.3.0-SNAPSHOT</version>
+    <version>1.3.2-SNAPSHOT</version>
   </parent>
 
   <artifactId>cxx-squid</artifactId>

--- a/integration-tests/pom.xml
+++ b/integration-tests/pom.xml
@@ -3,7 +3,7 @@
   <parent>
     <groupId>org.sonarsource.sonarqube-plugins.cxx</groupId>
     <artifactId>cxx</artifactId>
-    <version>1.3.0-SNAPSHOT</version>
+    <version>1.3.2-SNAPSHOT</version>
   </parent>
 
   <artifactId>integration-tests</artifactId>

--- a/pom.xml
+++ b/pom.xml
@@ -12,7 +12,7 @@
 
   <groupId>org.sonarsource.sonarqube-plugins.cxx</groupId>
   <artifactId>cxx</artifactId>
-  <version>1.3.0-SNAPSHOT</version>
+  <version>1.3.2-SNAPSHOT</version>
   <packaging>pom</packaging>
 
   <name>Cxx</name>
@@ -373,7 +373,7 @@
         <artifactId>guava</artifactId>
         <version>19.0</version>
       </dependency>
-    </dependencies>
+      </dependencies>
   </dependencyManagement>
   <build>
     <!-- To define the plugin version in your parent POM -->

--- a/sonar-c-plugin/pom.xml
+++ b/sonar-c-plugin/pom.xml
@@ -5,7 +5,7 @@
   <parent>
     <groupId>org.sonarsource.sonarqube-plugins.cxx</groupId>
     <artifactId>cxx</artifactId>
-    <version>1.3.0-SNAPSHOT</version>
+    <version>1.3.2-SNAPSHOT</version>
   </parent>
 
 

--- a/sonar-cxx-plugin/pom.xml
+++ b/sonar-cxx-plugin/pom.xml
@@ -5,7 +5,7 @@
   <parent>
     <groupId>org.sonarsource.sonarqube-plugins.cxx</groupId>
     <artifactId>cxx</artifactId>
-    <version>1.3.0-SNAPSHOT</version>
+    <version>1.3.2-SNAPSHOT</version>
   </parent>
 
   <artifactId>sonar-cxx-plugin</artifactId>

--- a/sslr-cxx-toolkit/pom.xml
+++ b/sslr-cxx-toolkit/pom.xml
@@ -4,7 +4,7 @@
   <parent>
     <groupId>org.sonarsource.sonarqube-plugins.cxx</groupId>
     <artifactId>cxx</artifactId>
-    <version>1.3.0-SNAPSHOT</version>
+    <version>1.3.2-SNAPSHOT</version>
   </parent>
 
   <artifactId>sslr-cxx-toolkit</artifactId>

